gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java

```diff
@@ -52,21 +52,16 @@ public DestinationDatasetHandlerService(SourceState jobState, Boolean canCleanUp
    * Executes handlers
    * @param workUnitStream
    */
-  public void executeHandlers(WorkUnitStream workUnitStream) {
-    if (handlers.size() > 0) {
-      if (workUnitStream.isSafeToMaterialize()) {
-        Collection<WorkUnit> workUnits = JobLauncherUtils.flattenWorkUnits(workUnitStream.getMaterializedWorkUnitCollection());
-          for (DestinationDatasetHandler handler : this.handlers) {
-            try {
-              handler.handle(workUnits);
-            } catch (IOException e) {
-              throw new RuntimeException(String.format("Handler %s failed to execute", handler.getClass().getName()), e);
-            }
-          }
-      } else {
-        throw new RuntimeException(DestinationDatasetHandlerService.class.getName() + " does not support work unit streams");
+  public WorkUnitStream executeHandlers(WorkUnitStream workUnitStream) {
+    for (DestinationDatasetHandler handler : this.handlers) {
+      try {
+       workUnitStream = handler.handle(workUnitStream);
```

Comment:
Can we use a new variable here to return? Even if it defaults to the parameter, not good practice to modify parameters directly